chain-signatures/node/src/mesh/connection.rs

```diff
@@ -1,245 +1,221 @@
-            self.connections.read().await.clone().keys_vec()
-        );
-    }
 
-    async fn set_participants(&self, participants: &Participants) {
-        *self.connections.write().await = participants.clone();
+        // drop the connections that are not in the seen list
+        self.drop_connections(seen);
     }
 
-    async fn set_potential_participants(&self, participants: &Participants) {
-        *self.potential_connections.write().await = participants.clone();
-        tracing::debug!(
-            "Pool set potential participants to {:?}",
-            self.potential_connections.read().await.keys_vec()
-        );
+    async fn connect_nodes(
+        &mut self,
+        participants: &Participants,
+        seen: &mut HashSet<Participant>,
+    ) {
+        for (participant, info) in participants.iter() {
+            seen.insert(*participant);
+
+            let node = (*participant, &info.url);
+            match self.connections.entry(*participant) {
+                Entry::Occupied(mut conn) => {
+                    if &conn.get().info != info {
+                        tracing::info!(target: "net[pool]", ?node, "node connection updating");
+                        conn.insert(NodeConnection::spawn(
+                            &self.client,
+                            participant,
+                            info,
+                            self.ping_interval,
+                        ));
+                    }
+                }
+                Entry::Vacant(conn) => {
+                    tracing::info!(target: "net[pool]", ?node, "node connection created");
+                    conn.insert(NodeConnection::spawn(
+                        &self.client,
+                        participant,
+                        info,
+                        self.ping_interval,
+                    ));
+                }
+            }
+        }
     }
 
-    pub async fn potential_participants(&self) -> Participants {
-        self.potential_connections.read().await.clone()
-    }
+    /// Drop connections that are not in the active connections list. Dropped connections
+    /// are no longer polled for their status.
+    fn drop_connections(&mut self, active_conn: HashSet<Participant>) {
+        let mut remove = Vec::new();
+        for participant in self.connections.keys() {
+            if !active_conn.contains(participant) {
+                remove.push(*participant);
+            }
+        }
 
-    async fn is_participant_stable(&self, participant: &Participant) -> bool {
-        self.status
-            .read()
-            .await
-            .get(participant)
-            .is_some_and(|state| match state {
-                StateView::Running { is_stable, .. } => *is_stable,
-                _ => false,
-            })
+        for participant in remove {
+            tracing::info!(target: "net[pool]", node = ?participant, "node connection dropping");
+            self.connections.remove(&participant);
+        }
```

Comment:
How about using [retain](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.retain)?

Something like this should work
```
self.connections.retain(|&participant, _connection|{
    active_conn.contains(participant)
})
```

Reply:
And I believe the logging here is currently be doubled by the `impl Drop for NodeConnection` logging anyway.